hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/namenode/FSDirRenameOp.java

```diff
@@ -470,17 +475,53 @@ static RenameResult unprotectedRenameTo(FSDirectory fsd,
       }
     } finally {
       if (undoRemoveSrc) {
-        tx.restoreSource();
+        tx.restoreSource(srcStoragePolicyCounts);
       }
       if (undoRemoveDst) { // Rename failed - restore dst
-        tx.restoreDst(bsps);
+        tx.restoreDst(bsps, dstStoragePolicyCounts);
       }
     }
     NameNode.stateChangeLog.warn("DIR* FSDirectory.unprotectedRenameTo: " +
         "failed to rename " + src + " to " + dst);
     throw new IOException("rename from " + src + " to " + dst + " failed.");
   }
 
+  /*
+   * Calculate QuotaCounts based on parent directory and storage policy
+   * 1. If the storage policy of src and dst are different,
+   *  calculate the QuotaCounts of src and dst respectively.
+   * 2. If all parent nodes of src and dst are not set with Quota,
+   *  there is no need to calculate QuotaCount.
+   * 3. if parent nodes of src and dst have Quota configured,
+   *  the QuotaCount is calculated once using the storage policy of src.
+   * */
+  private static void computeQuotaCounts(
+          QuotaCounts srcStoragePolicyCounts,
+          QuotaCounts dstStoragePolicyCounts,
+          INodesInPath srcIIP,
+          INodesInPath dstIIP,
+          BlockStoragePolicySuite bsps,
+          RenameOperation tx) {
+    INode dstParent = dstIIP.getINode(-2);
+    INode srcParentNode = FSDirectory.
+            getFirstSetQuotaParentNode(srcIIP);
+    INode srcInode = srcIIP.getLastINode();
+    INode dstParentNode = FSDirectory.
+            getFirstSetQuotaParentNode(dstIIP);
+    byte srcStoragePolicyID = FSDirectory.getStoragePolicyId(srcInode);
+    byte dstStoragePolicyID = FSDirectory.getStoragePolicyId(dstParent);
+    if (srcStoragePolicyID != dstStoragePolicyID) {
+      srcStoragePolicyCounts.add(srcIIP.getLastINode().
+              computeQuotaUsage(bsps));
+      dstStoragePolicyCounts.add(srcIIP.getLastINode()
+              .computeQuotaUsage(bsps, dstParent.getStoragePolicyID(), false,
+                      Snapshot.CURRENT_STATE_ID));
+    } else if (srcParentNode != dstParentNode || tx.withCount != null) {
+      srcStoragePolicyCounts.add(srcIIP.getLastINode().computeQuotaUsage(bsps));
+      dstStoragePolicyCounts.add(srcStoragePolicyCounts);
+    }
```

Comment:
what happen if `srcParentNode == dstParentNode && tx.withCount == null`?

Reply:
If this is the case, it can be understood that src and dst have a configured quota, and src is the isSrcInSnapshot attribute.

Reply:
Sorry I didn't get it. I means that we do nothing when `srcStoragePolicyID == dstStoragePolicyID && srcParentNode == dstParentNode && tx.withCount == null` because there is no `else` logic, So my question is what happen here, do we need to do something?
IMO, We don't need to compute quota usage once when `srcParentNode == dstParentNode`, right? However L524 seems to still compute it.

Reply:
I am confused with this condition `else if (srcParentNode == dstParentNode && tx.withCount == null){compute}` which is not `else if (srcParentNode == dstParentNode) {...}`, what's different between them here?

Reply:
@Hexiaoqiao @zhuxiangyi Our production cluster encountered performance issues related to rename and needed to be optimized.
In (srcStoragePolicyID == dstStoragePolicyID) condition, do we need to compute only when both conditions srcParentNode != dstParentNode and tx.isSrcInSnapshot == false are met? If it is right, can we use the below logics:
if (srcStoragePolicyID != dstStoragePolicyID) {
compute;
} else if (srcParentNode != dstParentNode && !tx.isSrcInSnapshot) {
compute;
}
Looking forward to your reply, thanks.